types/validator_set_test.go

```diff
@@ -658,6 +651,62 @@ type testVal struct {
 	power int64
 }
 
+func permutation(valList []testVal) []testVal {
+	if len(valList) == 0 {
+		return nil
+	}
+	permList := make([]testVal, len(valList))
+	perm := cmn.RandPerm(len(valList))
+	for i, v := range perm {
+		permList[v] = valList[i]
+	}
+	return permList
+}
+
+func createNewValidatorList(testValList []testVal) []*Validator {
+	valList := make([]*Validator, 0, len(testValList))
+	for _, val := range testValList {
+		valList = append(valList, newValidator([]byte(val.name), val.power))
+	}
+	return valList
+}
+
+func createNewValidatorSet(testValList []testVal) *ValidatorSet {
+	valList := createNewValidatorList(testValList)
+	valSet := NewValidatorSet(valList)
+	return valSet
+}
+
+func verifyValidatorSet(t *testing.T, valSet *ValidatorSet) {
+	// verify that the capacity and length of validators is the same
+	assert.Equal(t, len(valSet.Validators), cap(valSet.Validators))
+
+	// verify that the vals' tvp is set to the sum of the all vals voting powers
+	tvp := valSet.totalVotingPower
+	assert.Equal(t, valSet.TotalVotingPower(), tvp,
```

Comment:
I wonder if we actually test anything here. In order to test `is set to the sum of the all vals voting powers` we need to sum all vals powers, not use internal variable!

Reply:
we have to check valSet.totalVotingPower against the sum. I added a local function to check the sum since the TotalVotingPower() only computes when totalVotingPower is 0